src/lint/rules/variable-use-def.ts

```diff
@@ -0,0 +1,352 @@
+import type {
+  FragmentNode,
+  OrderedListNode,
+  UnorderedListNode,
+  TextNode,
+  UnderscoreNode,
+} from 'ecmarkdown';
+import type { Reporter } from '../algorithm-error-reporter-type';
+
+type HasLocation = { location: { start: { line: number; column: number } } };
+type VarKind =
+  | 'parameter'
+  | 'variable'
+  | 'abstract closure parameter'
+  | 'abstract closure capture'
+  | 'loop variable'
+  | 'attribute declaration';
+class Scope {
+  declare vars: Map<string, { kind: VarKind; used: boolean; node: HasLocation | null }>;
+  declare report: Reporter;
+  constructor(report: Reporter) {
+    this.vars = new Map();
+    // TODO remove this when regex state objects become less dumb
+    for (const name of ['captures', 'input', 'startIndex', 'endIndex']) {
+      this.declare(name, null);
+    }
+    this.report = report;
+  }
+
+  declared(name: string): boolean {
+    return this.vars.has(name);
+  }
+
+  // only call this for variables previously checked to be declared
+  used(name: string): boolean {
+    return this.vars.get(name)!.used;
+  }
+
+  declare(name: string, nameNode: HasLocation | null, kind: VarKind = 'variable'): void {
+    if (this.declared(name)) {
+      return;
+    }
+    this.vars.set(name, { kind, used: false, node: nameNode });
+  }
+
+  undeclare(name: string) {
+    this.vars.delete(name);
+  }
+
+  use(use: VariableNode) {
+    const name = use.contents[0].contents;
+    if (this.declared(name)) {
+      this.vars.get(name)!.used = true;
+    } else {
+      this.report({
+        ruleId: 'use-before-def',
+        message: `could not find a preceding declaration for ${JSON.stringify(name)}`,
+        line: use.location.start.line,
+        column: use.location.start.column,
+      });
+    }
+  }
+}
+
+type VariableNode = UnderscoreNode & { contents: [TextNode] };
+export function checkVariableUsage(
+  containingAlgorithm: Element,
+  steps: OrderedListNode,
+  report: Reporter
+) {
+  if (containingAlgorithm.hasAttribute('replaces-step')) {
+    // TODO someday lint these by doing the rewrite (conceptually)
```

Comment:
aww, I was looking forward to seeing how you did this

Reply:
It's not _that_ hard - you know exactly the step being replaced, so you can swap out that step in the original and re-run this algorithm, discarding any warnings which don't apply to the replacement section - but it just doesn't come up enough to be worth worrying about.